Update the student's modal to use the defined REST API pattern

gabrielcaires opened the pull request:
Related to #4959

### Changes proposed in this Pull Request
* It is changing the frontend to not use the `DELETE/PUT/PATCH` methods on destructive operations, due to issues with the old HTTP server that may block them.

* I remove all `http-client` tests related to nonce because now it is just an internal detail from the `apiFetch` lib.

* I am suggesting keeping the `http-client` abstraction in order to simplify possible changes 

### Extra notes
* After to add a student to a course, the modal is not closing automatically because the endpoint needs to return a value, today it is returning an empty body. It will be addressed in another PR. 
* The package json was using the wrong indentation size (2 spaces) when the project is configured to use 4 spaces, so this PR is also fixing it. 

### Testing instructions
* Use an HTTP server that blocks the `DELETE` method.
* Go to the students' page and remove a student from a course.

gabrielcaires committed:
Add unfetch polyfill to run test.

The jsom jest env by default doesn't have window.fetch implemented, so this polyfill is required to enable our components make requests inside the test context.

gabrielcaires committed:
Update http-client to use apiFetch instead of axios

We discovered we can't use the common http verbs on our endpoints due incompatibility with old http servers, so we back to use apiFetch.

It is adding a middleware simplify the usage of our API routes becase they need to be translated to a query_param.

donnapep commented:
> * The package json was using the wrong indentation size (2 spaces) when the project is configured to use 4 spaces, so this PR is also fixing it.

I think I'd prefer that we change the config to 2 spaces so I can fit more on one line.

gabrielcaires committed:
Fix incomplete mock

The test is breaking because it was setting the apiFetch module to be a simple function, when it is not exacly true. The apiFetch is function but also contains some methods like `.use`

gabrielcaires committed:
Reset nock

gabrielcaires committed:
Replace unfectch by whatwg-fetch because it is most used.

gabrielcaires committed:
Update github cache and checkout actions

gabrielcaires committed:
Clean nock before run every test

gabrielcaires committed:
Update npm engine

gabrielcaires commented on assets/admin/students/student-modal/course-list.test.js:
```diff
@@ -25,28 +25,26 @@ const courses = [
+	afterAll( () => nock.cleanAll() );
 
 	it( 'Should display courses in the list', async () => {
-		await act( async () => {
```
The `render` method already has an act, so it is not necessary.

gabrielcaires commented on assets/admin/students/student-modal/course-list.test.js:
```diff
@@ -83,16 +81,14 @@ describe( '<CourseList />', () => {
 		beforeEach( () => {
 			nock.cleanAll();
 			nock( 'http://localhost' )
-				.get( '/wp-json/wp/v2/courses' )
```
Using apiFake is not necessary to add the `wp-json` prefix.

gabrielcaires commented on assets/admin/students/student-modal/index.test.js:
```diff
@@ -36,17 +35,12 @@ describe( '<StudentModal />', () => {
-			.get( '/wp-json/wp/v2/courses' )
-			.query( { per_page: 100 } )
+			.get( '/wp/v2/courses' )
+			.query( { per_page: 100, _locale: 'user' } )
```
By default, api fetch add the `_locale` param.

gabrielcaires commented:
@donnapep  Regarding the tests failing, after a lot of tests, I found that [this commit](https://github.com/Automattic/sensei/pull/5029/commits/5b169a3f1da02d3459ff7fcf3cce545d23fb2584)  lead us to a more stable scenario.

merkushin commented:
> It is changing the frontend to not use the DELETE/PUT/PATCH methods on destructive operations, due to issues with the old HTTP server that may block them.

We decided to use these methods, didn't we?

donnapep reviewed (approved):
Works well, except for the documented issue with closing the modal after adding a student to a course.
